app/scripts/controllers/metametrics.js

```diff
@@ -87,6 +118,95 @@ export default class MetaMetricsController {
     );
   }
 
+  /**
+   * Create a funnel in state and returns the funnel object.
+   * @param {MetaMetricsFunnel} options - Funnel settings and properties
+   *  to initiate the funnel with.
+   * @returns {MetaMetricsFunnel}
+   */
+  createFunnel(options) {
+    const { funnels } = this.store.getState();
+
+    const id = generateUUID();
+    const funnel = {
+      id,
+      ...options,
+      lastUpdated: Date.now(),
+    };
+    this.store.updateState({
+      funnels: {
+        ...funnels,
+        [id]: funnel,
+      },
+    });
+    return funnel;
+  }
+
+  /**
+   * Updates a funnel in state
+   * @param {string} id - The funnel id to update
+   * @param {MetaMetricsFunnel} options - Funnel settings and properties
+   *  to initiate the funnel with.
+   * @returns {MetaMetricsFunnel}
+   */
+  updateFunnel(id, payload) {
```

Comment:
I do like the idea of having each fragment be a state machine, but I think that complexity is best tackled in a future PR that starts to make use of this underlying implementation. I agree right now the updateFragment method is a little too open-ended but I think hardening it will be best suited for when I start implementing events using it.